File: kernel/isr.c
#include <kernel/isr.h>
#include <kernel/pic.h>
#include <kernel/syscall.h>
#include <kernel/trapframe.h>

#include <stdint.h>
#include <stdio.h>

irq_handler_t irq_handlers[16] = { 0 };

void install_interrupt_handler(uint8_t irq, irq_handler_t handler) {
    irq_handlers[irq] = handler;
}

void uninstall_interrupt_handler(uint8_t irq) {
    irq_handlers[irq] = 0;
}

void interrupt_handler(uint8_t irq, trap_frame_t *trap_frame, uint32_t error) {
    if (irq < 32) {
        exception_handler(irq, error);
    } else if (irq == 48) {
        // syscall !!
        trap_frame->eax = syscall(trap_frame->eax, trap_frame->esp);
    } else {
        void (*handler)(uint8_t irq, trap_frame_t *trap_frame, uint32_t error) = irq_handlers[irq - 32];
        if (handler) {
            handler(irq, trap_frame, error);
        } else {
            kprintf("unhandled interrupt! irq: %d, error: 0x%x\n", irq, error);
        }
    }
    end_of_interrupt(irq);
}

void exception_handler(uint8_t irq, uint32_t error) {
    kprintf("fatal exception! irq: %d, error: 0x%x\n", irq, error);
    switch (irq) {
        case 14:
            // page fault
            uint32_t virtual_address;
            asm volatile ("movl %%cr2, %0" : "=a"(virtual_address));
            kprintf("page fault occurred while accessing virtual address 0x%x", virtual_address);
            break;

        default:
            break;
    }
}

File: kernel/src/isr.zig
const std = @import("std");
const pic = @import("pic.zig");
const stackframe = @import("stackframe.zig");
const writer = @import("serial.zig").writer;

const IrqHandler = ?*const fn () void;

var irq_handlers = std.mem.zeroes([16]IrqHandler);

pub fn install_handler(irq: u8, handler: IrqHandler) void {
    irq_handlers[irq] = handler;
}

pub fn uninstall_handler(irq: u8) void {
    irq_handlers[irq] = null;
}

export fn interrupt_handler(irq: u8, stack_frame: *stackframe.StackFrame, err: u32) void {
    _ = stack_frame;
    if (irq < 32) {
        exception_handler(irq, err);
    } else if (irq == 48) {
        // syscall
    } else {
        const handler = irq_handlers[irq - 32];
        if (handler) |handler_fn| {
            handler_fn();
        } else {
            writer.print("unhandled interrupt! irq: {}, error: {}\n", .{ irq, err }) catch unreachable;
        }
    }
    pic.end_of_interrupt(irq);
}

fn exception_handler(irq: u8, err: u32) void {
    writer.print("fatal exception! irq: {}, error: {}\n", .{ irq, err }) catch unreachable;
}
